### PR TITLE
Fix interface preserved by xml not preserving other implemented interfaces

### PR DESCRIPTION
**Pull request by mrvoorhe**

### CODE REVIEW

File: linker/Mono.Linker.Steps/MarkStep.cs [hunk 1]
```diff
@@ -1247,6 +1247,14 @@ protected virtual void MarkMethods (TypeDefinition type)
 		{
 			if (type.HasMethods)
 				MarkMethodCollection (type.Methods);
+
+			if (type.IsInterface) {
```
Comment (marek-safar) [inline linker/Mono.Linker.Steps/MarkStep.cs]: MarkMethod is not good place to do it as this is XML preserving only setting. I am also not sure if "all" should really mean all (e.g. custom attributes, constraint, base types, etc)

**Comment by TautvydasZilys** [comment]
This was a bug in the stripper that it wasn't preserving all interface methods. For instance, when marking IList, it would not mark GetEnumerator method. That seems wrong because any type that implements IList will have to implement GetEnumerator method.  Interfaces in IL are weird in the way that they can implement other interfaces, but they don't actually have the implemented interface methods on themselves. That's why if we are marking interface methods, we also have to mark the ones that will have to be implemented by delegation.

**Comment by marek-safar** [comment]
build

**Comment by marek-safar** [comment]
It seems to be that changing behaviour of `preserve=methods` to cover this special case is not great solution. I can certainly imagine scenario where I don't want to preserve base interface methods and in cases where member hides base member this is not deliberate.

Another problem shows when you use top level interface preserve setting to base interfaces. You can end up with preserve setting conflicts as base interface could have different preserve settings than interface inheriting from it.

I'd like to better understand WinRT scenario you are trying to fix with this. Could you share more code/context where this shows up?

**Comment by TautvydasZilys** [comment]
Certainly.

Windows Runtime has a concept of interface projection. That means that when a managed object implements a certain managed interface and then gets passed to native code, native code can assume that it implements the equivalent projected native interface and can call its methods.

For instance, interface System.Collections.Generic.IReadOnlyList`1<T> projects to Windows.Foundation.Collections.IVectorView`1<T>, while System.Collections.Generic.IEnumerable`1<T> projects to Windows.Foundation.Collections.IIterable`1<T>. System.Collections.Generic.IReadOnlyCollection`1<T> has no equivalent WIndows Runtime interface, so it doesn't project to anything.

Let's say a managed object implements IReadOnlyList`1<T>, and by delegation it also has to implement IReadOnlyCollection`1<T> and IEnumerable`1<T>. That means that COM Callable Wrapper for this managed object will have to implement Windows.Foundation.Collections.IVectorView`1<T> and Windows.Foundation.Collections.IIterable`1<T> interfaces. When you pass such object to native code, you have no way of knowing which interface methods will get called on them, so you need to play safe and assume that any of them gets called. In our fork of the linker we have additional mark step to find all managed objects that potentially implement COM/Windows Runtime interfaces and mark their methods (using MarkMethods).

Now, just marking IVectorView`1<T> and IIterable`1<T> methods are not enough. The managed objects don't actually implement these interfaces - they implement IReadOnlyList`1<T>, IReadOnlyCollection`1<T> and IEnumerable`1<T>. The runtime has to implement all IVectorView`1<T> and IIterable`1<T> methods in terms of calling the methods that were actually implemented by that object.

For example, let's take "T IVectorView`1<T>::GetAt(uint index)" method. The runtime will just be able to remap it to "T IReadOnlyList`1<T>::get_Item(int index)" method. That means to support this, we need to preserve that method. In general case, all the methods of managed interfaces that project to windows runtime interfaces if any non-stripped classes implement them. 

But that's not enough! Let's take another method: "uint IVectorView`1<T>::get_Size()". Even though IVectorView`1<T> is a projection of IReadOnlyList`1<T>, we can't implement it just using methods from IReadOnlyList`1<T>, as it has no way of querying its size. We have to go to an interface that was implemented by the managed object by delegation - IReadOnlyCollection`1<T>, and call "int IReadOnlyCollection`1<T>::get_Count()".

Therefore, we need to mark all implemented methods of interfaces that get implemented by interface implementation delegation.

**Comment by mmitche** [comment]
@dotnet-bot test this please

**Pull request closed without merging**